File: 048. Rotate Image/rotate.js
/**
 * @param {number[][]} matrix
 * @return {void} Do not return anything, modify matrix in-place instead.
 */
var rotate = function(matrix) {
    matrix.reverse()
    x = matrix[0].length
    y = matrix.length  
    for (var i=0;i<y;i++){
        for (var j=i;j<x;j++){ //just get upper triangle
            [matrix[j][i], matrix[i][j]] = [matrix[i][j], matrix[j][i]] //swap
        }
    }
};

// Runtime 92 ms / Memory 42.2 MB
// debug: https://jsfiddle.net/

File: 048. Rotate Image/rotate.cpp
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        int n = matrix.size();
        // vertical transfer
        reverse(matrix.begin(), matrix.end());
        // diagonal transfer
        for (int i=0; i<n; i++) {
            for (int j=i+1; j<n; j++) { //just get upper triangle
                swap(matrix[i][j],matrix[j][i]);
            }
        }
    }
};